### ISSUE TITLE
为什么我测试没有性能提升？

### ISSUE THREAD

**Issue by MeJerry215** [alipay/PainlessInferenceAcceleration#11]
测试脚本 改自 qwen_examples.py

测试4个句子的性能，反而是劣化的？将脚本中初始化模型的地方 互换一下，验证使用原生transformer 加载 qwen 以及使用pia加载qwen 后推理。

torch                         2.1.2

NVIDIA A100 80GB

nvcc --version
```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Wed_Sep_21_10:33:58_PDT_2022
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0
```

```python
"""
# -*- coding: utf-8 -*-
"""
Copyright (c) Ant Financial Service Group and its affiliates.
"""

from transformers import AutoModelForCausalLM
import os
import sys
import time
import torch
from transformers import AutoTokenizer
from transformers.generation import GenerationConfig

from pia.lookahead.common.pretrained_model import LookaheadCache
from pia.lookahead.models.qwen.modeling_qwen import QWenLMHeadModel
from pia.lookahead.models.qwen.tokenization_qwen import QWenTokenizer
from pia.lookahead.examples import local_path_dict

# model_dir = local_path_dict.get('qwen', 'your/model/path')
model_dir = "llm_models/Qwen-7B-Chat"
dtype = torch.float16 if torch.cuda.is_available() else torch.float32

# model = QWenLMHeadModel.from_pretrained(model_dir  # , cache_dir='../'
#                                         , torch_dtype=dtype, fp16=True, low_cpu_mem_usage=True, device_map={"": "cuda:0"}
#                                         ).cuda().eval()


model = AutoModelForCausalLM.from_pretrained(model_dir,
                                             torch_dtype=dtype,
                                             fp16=True,
                                             low_cpu_mem_usage=True,
                                             trust_remote_code=True,
                                             device_map={"":"cuda:0"},).cuda().eval()
model.generation_config = GenerationConfig.from_pretrained(model_dir, trust_remote_code=True)


tokenizer = QWenTokenizer.from_pretrained(model_dir)
stop_words = [tokenizer.encode(x)[0] for x in [',', '.', ' ', '，', '。']]


# prompt = "杭州在哪里？"

prompts = [
    "编一个200字左右的儿童故事",
    "描述一下中国的地理位置和环境信息",
    "从南京如何去云南大理?",
    "给我写一首关于赞美春天的诗？"
]

device = 'cuda:0' if torch.cuda.is_available() else 'cpu'
i = 0
history = None

warmup_prompt = "你好"
debug_lookahead = False
decoding_length = 64
branch_length = 12
max_new_tokens = 256
use_lookahead= True
decoding_kwargs = {"use_lookahead": use_lookahead,
                   "debug_lookahead": debug_lookahead,
                   "decoding_mode": 'hier',
                   "decoding_length": decoding_length,
                   "branch_length": branch_length,
                   "stop_words": stop_words}
model.generation_config.decoding_kwargs = decoding_kwargs
model.generation_config.do_sample = False  # default is True for qwen, result in different responses in every generation

for i in range(10):
    response, history = model.chat(tokenizer, warmup_prompt, history=None, eos_token_id=151645)

print(response)

for prompt in prompts:
    # repetition_penalty is not fully supported currently, will fix in the future
    model.generation_config.repetition_penalty = None
    ts = time.time()
    response, history = model.chat(tokenizer, prompt, history=None, eos_token_id=151645)
    te = time.time()
    token_count = len(tokenizer.encode(response))
    print(f'lookahead:{use_lookahead} time:{te - ts:.3f}s speed:{token_count/(te-ts):.1f}token/s response:{response}\n\n\n')

```

左为transformer  右为pia

![image](https://github.com/alipay/PainlessInferenceAcceleration/assets/53092165/d2264dac-d969-4c02-8db3-5688fc6b7386)

而如果我把prompt 全都换成 编一个200字左右的儿童故事，性能有很大的提升，但是我想请问一下，就是你们业务部署的时候很容易遇到这种场景？或者说 pia 的应用场景是什么？

![image](https://github.com/alipay/PainlessInferenceAcceleration/assets/53092165/90029756-9d29-4cea-bbb1-6514fece1778)

然后在没有sample的情况下，推理4次相同的prompt，产生了3种不同的case？
![image](https://github.com/alipay/PainlessInferenceAcceleration/assets/53092165/aef4e74c-4d16-46be-a45c-94883000b19a)

还是说我测试的方法不太对？

**Comment by ZipECHO** [comment]
论文中使用的方法是使用大量的数据进行warm up，在这种情况下会有所提升

**Comment by MeJerry215** [comment]
> 论文中使用的方法是使用大量的数据进行warm up，在这种情况下会有所提升

大量是多少，我现在整在用webqa test数据集进行warmup，3000+ examples，transformer的性能大约是42 tokens/s

参考数据集 链接 https://kexue.fm/archives/4338

warmup 使用的是  me_test.ir.json 3024 个questions

test的使用的时 从me_validation.json 抽取出来的questions

现在重新改了一下脚本，如下

```
from transformers import AutoModelForCausalLM, AutoTokenizer
import os
import sys
import time
import torch
import pdb
import json

from pia.lookahead.common.pretrained_model import LookaheadCache
from pia.lookahead.models.qwen.modeling_qwen import QWenLMHeadModel
from pia.lookahead.models.qwen.tokenization_qwen import QWenTokenizer
from transformers.generation import GenerationConfig
from tqdm import tqdm


model_dir = "Qwen-7B-Chat"

dtype = torch.float16
use_pia = True
tokenizer = QWenTokenizer.from_pretrained(model_dir)
stop_words = [tokenizer.encode(x)[0] for x in [',', '.', ' ', '，', '。']]


with open("./webqa/me_test.ir.json", "r") as f:
    test = json.load(f)

with open("./webqa/me_validation.ir.json", "r") as f:
    validation = json.load(f)


# with open("test.json", "w", encoding="utf-8") as f:
#     json.dump(test, f, ensure_ascii=False)


# with open("validation.json", "w", encoding="utf-8") as f:
#     json.dump(validation, f, ensure_ascii=False)

questions = []

for key, val in test.items():
    questions.append(val['question'])

validation_questions = []
for key, val in validation.items():
    validation_questions.append(key)

if use_pia:
    model = QWenLMHeadModel.from_pretrained(model_dir  # , cache_dir='../'
                                            , torch_dtype=dtype, fp16=True, low_cpu_mem_usage=True, device_map={"": "cuda:0"}
                                            ).cuda().eval()
    model.generation_config = GenerationConfig.from_pretrained(model_dir, trust_remote_code=True)
    debug_lookahead = False
    decoding_length = 64
    branch_length = 12
    max_new_tokens = 256
    use_lookahead = True
    decoding_kwargs = {"use_lookahead": use_lookahead,
                       "debug_lookahead": debug_lookahead,
                       "decoding_mode": 'hier',
                       "decoding_length": decoding_length,
                       "branch_length": branch_length,
                       "stop_words": stop_words}
    model.generation_config.decoding_kwargs = decoding_kwargs
    model.generation_config.do_sample = False
    model.generation_config.repetition_penalty = None

else:
    model = AutoModelForCausalLM.from_pretrained(model_dir,
                                                 torch_dtype=dtype,
                                                 fp16=True,
                                                 low_cpu_mem_usage=True,
                                                 trust_remote_code=True,
                                                 device_map={"": "cuda:0"},).cuda().eval()


def infer(prompt):
    ts = time.time()
    response, history = model.chat(tokenizer, prompt, history=None, eos_token_id=151645)
    te = time.time()
    token_count = len(tokenizer.encode(response))
    print(f"prompt: {prompt} response: {response} time: {te- ts:.3f} seconds speed: {token_count/(te-ts):.1f} token/s\n\n")


# warm up:
for question in tqdm(questions[:]):
    infer(question)

pdb.set_trace()
print("============================> start validation <============================")
print("============================> start validation <============================")
print("============================> start validation <============================")
print("============================> start validation <============================")


for question in tqdm(validation_questions[:100]):
    infer(question)
```

基线性能是使用transformer 进行推理时的42 tokens/s  和遇到是什么句子没有关系。

这样进行warmup 应该是对的吧

**Comment by chenliangjyj** [comment]
可以参考下论文看下加速的具体原理， full bechmark在 [连接](https://github.com/alipay/PainlessInferenceAcceleration/tree/main/pia/lookahead/benchmarks)

对于四次生成不一样的原因，我们发现qwen在推断时对精度比较敏感，在fp16的情况下，建议开启 repetition_penalty， 使用如下脚本生成是一致的。

``` python
import os
import sys
import time
import torch
from transformers import AutoTokenizer
from transformers.generation import GenerationConfig

from pia.lookahead.models.qwen.modeling_qwen import QWenLMHeadModel
from pia.lookahead.models.qwen.tokenization_qwen import QWenTokenizer
from pia.lookahead.examples import local_path_dict

model_dir = local_path_dict.get('qwen', 'your/model/path')

dtype = torch.float16 if torch.cuda.is_available() else torch.float32
device = 'cuda:0' if torch.cuda.is_available() else 'cpu'
model = QWenLMHeadModel.from_pretrained(model_dir
                                       , cache_dir='../'
                                       , torch_dtype=dtype
                                       , fp32=True
                                       , low_cpu_mem_usage=True
                                       , device_map={"": device}
                                       ).half().cuda().eval()
model.generation_config = GenerationConfig.from_pretrained(model_dir, trust_remote_code=True)

tokenizer = QWenTokenizer.from_pretrained(model_dir)
stop_words = [tokenizer.encode(x)[0] for x in [',', '.', ' ', '，','。']]


# prompt = "杭州在哪里？"
prompt = "编一个200字左右的儿童故事"

for use_lookahead in [False, True, True]:
    decoding_length = 64
    branch_length = 12
    debug_lookahead = False
    max_new_tokens = 256
    decoding_kwargs = {"use_lookahead": use_lookahead,
                       "debug_lookahead": debug_lookahead,
                       "decoding_length": decoding_length,
                       "branch_length": branch_length,
                       "stop_words": stop_words,
                       "tokenizer": tokenizer}
    model.generation_config.decoding_kwargs=decoding_kwargs
    model.generation_config.repetition_penalty = 1.5
    model.generation_config.do_sample=False  # default is True for qwen, result in different responses in every generation
    ts = time.time()
    response, history = model.chat(tokenizer, prompt, history=None, eos_token_id=151645)
    te = time.time()
    token_count = len(tokenizer.encode(response))
    print(f'lookahead:{use_lookahead} time:{te - ts:.3f}s speed:{token_count/(te-ts):.1f}token/s response:\n{response}\n')

```
